caching: Fully parse and handle request and response cache-control headers

yosrym93 opened the pull request:
Commit Message: CacheFilter fully parses and handles request and response cache-control headers

Additional Description:
Request and response cache-control headers are fully parsed in `CacheHeadersUtils`. `isCacheableResponse` in the `CacheFilter` now uses parsed cache-control headers to check if a received response can be cached. `HttpCache` now uses parsed cache-control headers to check if a cached response requires validation before it is served. Cache entries validation is not yet implemented.
Signed-off-by: Yosry Ahmed <yosryahmed@google.com>

Risk Level: Low
Testing: Unit testing
Docs Changes: N/A
Release Notes: N/A
Fixes #9833

capoferro commented on source/extensions/filters/http/cache/http_cache_utils.h:
```diff
@@ -1,34 +0,0 @@
-#pragma once
```
You can do a `git mv` to preserve the change history when renaming files as well as making it possible to see what's changed in the files. It's also helpful for clarity in the review to rename things in a commit separate from any functionality change (same PR, just a separate commit) to make the resulting diff as clean as it can be.

yosrym93 replied:
Thanks for the tip! In this case I rewrote the whole file, the only thing remaining from http_cache_utils is the `httpTime` function. So I do not think it should matter for this specific case. However, is there a way to revert these changes and separate them into different commits?

yosrym93 replied:
After a bit of reading I figured `git mv` will preserve the change history and the file rename in case the changes in the file are minor enough that git can keep track of them. I should have had a separate commit to rename the file like you mentioned.

capoferro commented on source/extensions/filters/http/cache/cache_headers_parser.h:
```diff
@@ -0,0 +1,102 @@
+
+// std::ostream& operator<<(std::ostream& os, const ResponseCacheControl& response_cache_control);
+
+class CacheHeadersParser {
```
The name `CacheHeadersParser` makes me think you're intended to instantiate a parser object and use it to do things, rather than it being a collection of helper functions. Envoy naming convention seems to be to name a collection of static functions a "utility", for example, https://github.com/envoyproxy/envoy/blob/master/source/common/http/path_utility.h

yosrym93 replied:
Makes sense, will modify it. Do you think I should merge this class with the existing `CacheFilterUtils` class, or keep them separate?

toddmgreer commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,157 @@
+
+std::tuple<absl::string_view, absl::string_view>
+CacheHeadersUtils::separateDirectiveAndArgument(absl::string_view full_directive) {
+  std::vector<absl::string_view> directive_parts =
```
Whitespace is only allowed between full directives; it would be simpler to only StripAsciiWhitespace once (on full_directive). Also, if you change the return type of this function from tuple to pair, I think you can simplify the function body to:
{
  return absl::StrSplit(absl::StripAsciiWhitespace(full_directive), absl::MaxSplits('=', 1));
}

yosrym93 replied:
Thanks for letting me know. Will do.

yosrym93 replied:
I did not know absl::StrSplit could return an std::pair, great to know!

toddmgreer commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,157 @@
+
+RequestCacheControl CacheHeadersUtils::requestCacheControl(absl::string_view cache_control_header) {
+  RequestCacheControl request_cache_control;
+  std::vector<absl::string_view> directives = absl::StrSplit(cache_control_header, ',');
```
How well does this code handle request directives with ',' in the argument? If the worst that can happen is an incorrect result, that should be ok, but please add a comment warning future maintainers to be careful.

yosrym93 replied:
The worst is that this directive could be interpreted in a wrong way.
For example: `cache-control: private, max-age=5, max-stale=10,0, no-store`.
This will split to:
`private` , `max-age=5` , `max-stale=10`, `0` , `no-store`
The `0` will be ignored, and `max-stale` will be misinterpreted as 10.

yosrym93 replied:
Which kinda make sense. 
A header like this is probably more likely to be intended as `max-stale=10` and another directive that was written incorrectly as `0`, rather than `max-stale=100` with a random injected comma.

yosrym93 replied:
Added comment with an example.

yosrym93 replied:
@capoferro it's not an intended behavior, but rather an output of a wrong input, which shouldn't matter. I mean if the behavior changes to max-stale not being parsed at all for example, that should not be a problem (as the input is ill-formed anyway). What do you think?

capoferro replied:
Ensuring that we handle invalid input is also worthy of a test. Tests serve as living documentation of the intended behavior of a function.

yosrym93 replied:
I already have some test cases of invalid inputs, will add one for this case.

toddmgreer commented on source/extensions/filters/http/cache/http_cache.cc:
```diff
@@ -68,29 +70,50 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
+
+  if (timestamp_ > expiration_time) {
+    // Response is stale, return false unless max-stale value is higher than overdue expiration
+    return !(!response_cache_control.no_stale && request_cache_control_.max_stale.has_value() &&
```
Please distribute the outer '!' inside the parens

toddmgreer commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,157 @@
+#include "extensions/filters/http/cache/cache_headers_utils.h"
+
+#include <bits/stdint-intn.h>
```
Why are these 'bits' headers needed? Are they portable?

yosrym93 replied:
I did not add these, probably vscode extension, not the first time it did that but I did not notice these.
Will revise all the includes.

yosrym93 commented on include/envoy/http/header_map.h:
```diff
@@ -315,6 +315,7 @@ class HeaderEntry {
   HEADER_FUNC(EnvoyUpstreamCanary)                                                                 \
   HEADER_FUNC(EnvoyUpstreamHealthCheckedCluster)                                                   \
   HEADER_FUNC(EnvoyUpstreamServiceTime)                                                            \
+  HEADER_FUNC(Expires)                                                                             \
```
Will do.

yosrym93 commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -60,7 +63,9 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
 Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
                                                      bool end_stream) {
-  if (lookup_ && CacheFilterUtils::isCacheableResponse(headers)) {
+  // lookup_ is nullptr (false) if the request was non-cache-able,
```
Will change it.

yosrym93 commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,107 @@
+
+struct RequestCacheControl {
+  // must_validate is true if 'no-cache' directive is present
+  bool must_validate = false;
```
Will do.

yosrym93 commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,107 @@
+  OptionalDuration min_fresh;
+  OptionalDuration max_stale;
+
+  bool operator==(const RequestCacheControl& rhs) const {
```
Will refactor them out of the structs.

yosrym93 commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,107 @@
+  // A directive that should not have arguments will ignore any existing arguments
+  // A directive that should have arguments will be ignored if no arguments are present
+  // Unknown directives are ignored
+  static ResponseCacheControl responseCacheControl(absl::string_view cache_control_header);
```
Yeah makes more sense, will change them.

yosrym93 commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,107 @@
+  // Unknown directives are ignored
+  static ResponseCacheControl responseCacheControl(absl::string_view cache_control_header);
+
+private:
```
I don't understand what you mean. Can you elaborate?

yosrym93 replied:
Do you mean I should just define them as global functions in the .cc file?

toddmgreer replied:
Declare/define them in an unnamed namespace in the .cc file. There's no value in putting them in the .h file--why should files that include this .h file include declarations for functions they can't call?

yosrym93 commented on source/extensions/filters/http/cache/cacheability_utils.cc:
```diff
@@ -0,0 +1,50 @@
+  // https://tools.ietf.org/html/rfc7231#section-6.1,
+  // https://tools.ietf.org/html/rfc7538#section-3,
+  // https://tools.ietf.org/html/rfc7725#section-3
+  // TODO: the list of cache-able status codes should be configurable
```
I tried that, but the spellchecker wouldn't let me. Should I add it to the dictionary or whitelist files in tools/spelling?

yosrym93 replied:
I found cacheability in the dictionary so I added cacheable, will fix the comments.

yosrym93 commented:
@toddmgreer I addressed all your comments except those related to `CacheHeaderUtils`, waiting for further explanation on the private members comment.
Looks like this class will be heavily refactored or removed.

yosrym93 commented on source/extensions/filters/http/cache/http_cache.cc:
```diff
@@ -68,29 +70,54 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
+
+  if (response_cache_control.must_validate || request_cache_control_.must_validate ||
+      (request_cache_control_.max_age.has_value() &&
+       request_cache_control_.max_age.value() > response_age)) {
```
It checks multiple (somehow) unrelated things. It checks if the response or request explicitly require revalidation or if the request's max-age condition requires revalidation. It could be extracted to 2 or 3 functions but then I think that defeats the purpose?

capoferro replied:
Something like this can be clearer (perhaps the names could be better, but something like this):

`if (mustValidate(response_cache_control) || !maxAgeIsSatisfied(response_age))`

capoferro replied:
Or perhaps just encapsulate the second bit:

`response_cache_control.must_validate || request_cache_control_.must_validate || !maxAgeIsSatisfied(response_age)`

yosrym93 replied:
Maybe I can add similar ones for `min-fresh` and `max-stale` too, but with inverse logic (as that's what is always checked).
Something like
```
bool maxAgeViolated(response_age);
bool minFreshViolated(current_time, expiration_time);
bool maxStaleViolated(current_time, expiration_time);
```

capoferro commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,160 @@
+  return duration.has_value() ? os << duration.value().count() : os << " ";
+}
+
+// TODO: clean or remove
```
Can we resolve this TODO before merge or should we make a ticket to track the work we need to do?

yosrym93 replied:
Oh I already did, but forgot to remove the TODOs. Will remove them.

yosrym93 commented:
@toddmgreer all comments resolved.

yosrym93 commented:
/azp run envoy-presubmit

mpwarres commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,182 @@
+OptionalDuration parseDuration(absl::string_view s) {
+  OptionalDuration duration;
+  // Strip quotation marks if any
+  if (!s.empty() && s.front() == '"' && s.back() == '"') {
```
What happens if s is (somehow) a string_view of length 1 containing the single character '"'? Seems like this might then call substr with a negative length.

yosrym93 replied:
abseil handles this internally, if `s = "` , s.substr(1, -1) is an empty string

yosrym93 replied:
However, I am not sure if that's a defined behavior that we can count on. Probably worth a check.
Thanks!

jmarantz commented:
@toddmgreer can you take a pass? Thanks.

yosrym93 commented:
@jmarantz any updates on this?

jmarantz commented:
Was hoping @toddmgreer could take a pass first. Todd; gentle ping?

toddmgreer commented:
Sorry for the delay--much craziness. I'll go through this tomorrow
(Friday).

On Thu, Jul 9, 2020 at 6:12 AM Joshua Marantz <notifications@github.com>
wrote:

> Was hoping @toddmgreer <https://github.com/toddmgreer> could take a pass
> first. Todd; gentle ping?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/11727#issuecomment-656118828>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFRAWPJQMK5TWBDDBPN6E2LR2W65TANCNFSM4OGJPUHA>
> .
>

toddmgreer commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -37,13 +37,17 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   ASSERT(decoder_callbacks_);
-  lookup_ = cache_.makeLookupContext(LookupRequest(headers, time_source_.systemTime()));
+
+  auto lookup_request = LookupRequest(headers, time_source_.systemTime());
```
No need for auto here:
LookupRequest lookup_request(headers, time_source_.systemTime());

yosrym93 replied:
Edited.

toddmgreer commented on source/extensions/filters/http/cache/cacheability_utils.h:
```diff
@@ -0,0 +1,33 @@
+  static bool isCacheableResponse(const Http::ResponseHeaderMap& headers);
+
+private:
+  const static absl::flat_hash_set<absl::string_view> cacheable_status_codes_;
```
Please remove the declaration of cacheaable_status_codes_ from this header file, remove "CacheabilityUtils::" from its definition in the .cc file, put it in an unnamed namespace, and rename it to follow Envoy style rules for non-member constants.

yosrym93 replied:
Done. 
Just to make sure I understand the principle correctly, any private members/functions that do not use any members of the class are better replaced by variables/functions in an unnamed namespace in the .cc file. Correct?

toddmgreer replied:
Yes. There are two reasons to put a declaration in a .h file:
1. The language requires it.
2. It's needed by more than one file.

Otherwise, there's no reason to clutter up the header.

yosrym93 replied:
Thanks for the clarification.

toddmgreer commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,71 @@
+using OptionalDuration = absl::optional<SystemTime::duration>;
+
+struct RequestCacheControl {
+  RequestCacheControl(bool must_validate = false, bool no_store = false, bool no_transform = false,
```
I think the code would be simpler if RequestCacheControl and ResponseCacheControl's only constructors were the string_view ones. In tests, they can be constructed from string literals (e.g. "public, max-age=34").

yosrym93 replied:
In tests, I am checking that a string cache-control header is parsed correctly to booleans and durations, how can I test this if I use a string literals?

yosrym93 replied:
I could create a `TestRequestCacheControl` struct in the test file that inherits `RequestCacheControl` and defines that constructor. Same for `ResponseCacheControl`. What do you think?

yosrym93 replied:
This would require adding a default constructor to `RequestCacheControl` and `ResponseCacheControl` though, it doesn't have to do anything but it must be defined. Also, I would not be able to initialize the members in the initializer list of the derived test classes, but rather in the constructor itself. It will be something like this:
```
struct TestRequestCacheControl : public RequestCacheControl {
  TestRequestCacheControl(bool must_validate, bool no_store, bool no_transform, bool only_if_cached,
                          OptionalDuration max_age, OptionalDuration min_fresh,
                          OptionalDuration max_stale) {
    must_validate_ = must_validate;
    no_store_ = no_store;
    no_transform_ = no_transform;
    only_if_cached_ = only_if_cached;
    max_age_ = max_age;
    min_fresh_ = min_fresh;
    max_stale_ = max_stale;
  }
};
```

toddmgreer replied:
That works, and avoids complicating the public interface. (Ignore the string literal suggestion--I was erroneously thinking about tests for code that uses these structs.)

toddmgreer commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,71 @@
+                      bool only_if_cached = false, OptionalDuration max_age = OptionalDuration(),
+                      OptionalDuration min_fresh = OptionalDuration(),
+                      OptionalDuration max_stale = OptionalDuration());
+  RequestCacheControl(absl::string_view cache_control_header);
```
One-argument constructors should generally be declared 'explicit'.

toddmgreer commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,182 @@
+}
+
+std::ostream& operator<<(std::ostream& os, const RequestCacheControl& request_cache_control) {
+  return os << "{"
```
Keep them this way if you like, but please consider instead only reporting the "true" bits (e.g. "no-store, min_fresh=12")

yosrym93 replied:
After the cache validation PR I will dedicate a PR to cleaning and improving tests and coverage. One of my todos for this is to move these operators to the test files (as they are only used for testing) and clean them. I would rather postpone this for that PR.

yosrym93 replied:
Also the presence of these operators in the source code makes the coverage appear to be lower than it actually is, as these operators are only invoked on test failures.

toddmgreer replied:
SGTM

toddmgreer commented on test/extensions/filters/http/cache/cache_headers_utils_test.cc:
```diff
@@ -0,0 +1,263 @@
+namespace Cache {
+namespace {
+
+#define DURATION(s) std::chrono::seconds(s)
```
Please delete these macros. In this codebase, it's preferred to spell things out in every use.

yosrym93 replied:
I only use them in this test file and I think without them the test cases will look way bigger, uglier, and less clear. Don't you agree?

toddmgreer replied:
It doesn't matter whether I agree. These macros won't get accepted by the Envoy maintainers.

yosrym93 replied:
I can't see anything in the style guide against it, I'd rather take a shot at it with the maintainers.

toddmgreer replied:
@jmarantz, your call.

jmarantz replied:
Envoy style allows macros but these looks like they could be inline functions or constants. If you think they add a ton of value I'd recommend using those in general. For these cases:

IMO `DURATION(s)` obscures what's happening and what it replaces is not much longer.

For `UNSET_DURATION` you should just use `absl::nullopt`

`MAX_DURATION` .... again I think what it replaces is not too long and is clearer.

So I'm leaning against them, not because they are against style guide but because I think the code will be easier to read without the indirection.

toddmgreer commented on test/extensions/filters/http/cache/cache_headers_utils_test.cc:
```diff
@@ -0,0 +1,263 @@
+class RequestCacheControlTest : public testing::TestWithParam<RequestCacheControlTestCase> {
+public:
+  // C++11 static members must be initialized out of line
+  const static RequestCacheControlTestCase test_cases[];
```
Non-PoD static/global variables are forbidden. The Envoy style guide encourages following the advice at https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use. You can also look at other uses of INSTANTIATE_TEST_SUITE_P and copy their approaches.

yosrym93 replied:
I will replace them with static functions returning the test cases.

toddmgreer commented on source/extensions/filters/http/cache/cacheability_utils.cc:
```diff
@@ -0,0 +1,55 @@
+// https://tools.ietf.org/html/rfc7538#section-3,
+// https://tools.ietf.org/html/rfc7725#section-3
+// TODO(yosrym93): the list of cacheable status codes should be configurable
+const static absl::flat_hash_set<absl::string_view> CacheableStatusCodes = {
```
non-POD static variables aren't allowed. In this case, I suggest replacing the use of CacheableStatusCodes with parsing the number and using a switch statement.

yosrym93 replied:
I can create a static function `isCacheableStatusCode` and define the static `absl::flas_hash_set` inside it, right?

toddmgreer replied:
yes, inside a function, you can have:
static absl::flash_hash_set& codes = *new absl::flas_hash_set(...

yosrym93 commented:
@toddmgreer I think I covered your comments, take a look whenever you can.

toddmgreer commented:
@jmarantz This concludes my final round of reviews for this PR. Please take a look.

yosrym93 commented:
@jmarantz do you need anything else before the review?

jmarantz commented:
I'm going to take a quick look. One quick note: please avoid force-pushing into Envoy github going forward; it makes reviews very difficult.

yosrym93 commented:
> I'm going to take a quick look. One quick note: please avoid force-pushing into Envoy github going forward; it makes reviews very difficult.

Thanks a lot! and sorry about force pushing.

jmarantz commented on source/extensions/filters/http/cache/cache_filter.h:
```diff
@@ -56,6 +57,8 @@ class CacheFilter : public Http::PassThroughFilter,
   // TODO(toddmgreer): cache trailers.
   bool response_has_trailers_;
 
+  bool request_allows_inserts_;
```
please init this here or in ctor. Also the bool above.

jmarantz commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -37,13 +37,17 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
-  lookup_ = cache_.makeLookupContext(LookupRequest(headers, time_source_.systemTime()));
+
+  LookupRequest lookup_request(headers, time_source_.systemTime());
+  request_allows_inserts_ = !lookup_request.requestCacheControl().no_store_;
```
suggest leaving link to http caching spec describing the semantics; I'm not sure if absence of "no-store" on request is all that we need to know we haven't disallowed saving cache from the request. It probably is but a spec ref would be awesome.

yosrym93 replied:
I will add the relevant RFC link at the bool definition.

jmarantz commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,170 @@
+
+// Utility functions used in RequestCacheControl & ResponseCacheControl
+namespace {
+OptionalDuration parseDuration(absl::string_view s) {
```
suggest making these less private so you can unit-test them directly.

yosrym93 replied:
For now, they are only used by and tested through RequestCacheControl and ResponseCacheControl, that's why they are private to this file.

yosrym93 replied:
@toddmgreer what do you think?

jmarantz replied:
Seems like it would be easier to thoroughly test this on its own.

toddmgreer replied:
I'm not a fan of making helpers public for testing when (as is the case here) we can adequately test the public API instead. However, I'll support whatever you work out with @jmarantz.

yosrym93 replied:
@jmarantz I am happy to add tests for the function, my point is just that it is fairly straight forward and is already tested through existing tests for RequestCacheControl & ResponseCacheControl.

jmarantz replied:
I believe the positive cases are all tested. I was thinking of testing all kinds of malformed durations. Are they all tested? They might be but it's harder to tell cause there's no focused testing at this layer.

yosrym93 replied:
The only logic this function performs is removing any existing quotes (which is tested). The actual parsing and detection of malformed digits is actually done through `absl::SimpleAtoi`. If it parses a valid number, it is used as the duration, otherwise the duration is ignored.

yosrym93 replied:
Some malformed durations are tested through `RequestCacheControl` and `ResponseCacheControl`, but not all kinds of malformed durations as this would really be testing of `absl::SimpleAtoi`.

jmarantz commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,170 @@
+// VCHAR           =  %x21-7E  ; visible (printing) characters
+
+RequestCacheControl::RequestCacheControl(absl::string_view cache_control_header) {
+  must_validate_ = no_store_ = no_transform_ = only_if_cached_ = false;
```
init these in the declaration.

yosrym93 replied:
Member initialization of bitfields in declarations is a C++20 feature.

yosrym93 replied:
If these are replaced by booleans I will initialize them in the declaration

jmarantz commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,170 @@
+  long num;
+  if (absl::SimpleAtoi(s, &num) && num >= 0) {
+    // s is a valid string of digits representing a positive number
+    duration = std::chrono::seconds(num);
```
should we have some kind of error-state return so that we can drop the entire request if the duration is malformed?

yosrym93 replied:
The current behavior is that directives with a malformed duration are ignored. The RFC [https://httpwg.org/specs/rfc7234.html#delta-seconds] does not specify what should be done if a duration is malformed.

jmarantz replied:
Comments referencing the RFC when making policy decisions would be helpful.

jmarantz commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,66 @@
+  explicit ResponseCacheControl(absl::string_view cache_control_header);
+
+  // must_validate is true if 'no-cache' directive is present; arguments are ignored for now
+  bool must_validate_ : 1;
```
is there a compelling reason to bit-pack these? You have 5 bools here which would easily fit into a uint64_t, which I think you'll pay for anyway due to alignment.

yosrym93 replied:
It was a suggestion of @toddmgreer 's review.

jmarantz replied:
Todd what's the advantage of bit-packing these? At first glance it's (a) not smaller due to alignment and (b)  this is a temp structure and it probably doesn't matter how small it is.

yosrym93 replied:
The `ResponseCacheControl` struct is indeed only used temporarily in one function for now. However, the `RequestCacheControl` struct is a member of `LookupRequest` that lives for a bit longer (for the duration of the cache lookup).

jmarantz replied:
Makes sense. Still I want to know what bit-packing buys in terms of overall sizeof this struct. It almost certainly costs you instructions, as well as the ability to nit in the declaration.

toddmgreer replied:
Apparently, I either can't count, or hallucinated additional member variables that pushed them past 56 and 24 bytes. I rechecked, and indeed the bit-packing has no benefits:
RequestCacheControl is 56 bytes either way, and ResponseCacheControl is 24 bytes either way. Ditch it.

jmarantz commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,170 @@
+  // Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate
+  // Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format
+  // Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
+  static const auto& rfc7231_date_formats = *new std::array<std::string, 3>{
```
use CONSTRUCT_ON_FIRST_USE macro from source/common/common/macros.h.

actually you can do this with a pod:
```
static const char* rfc7231_date_formats[] = {
  "%a, %d %b %Y %H:%M:%S GMT",
  "%A, %d-%b-%y %H:%M:%S GMT",
  "%a %b %e %H:%M:%S %Y"
};
```

yosrym93 replied:
Actually this function does not belong to the PR it was in the deleted http_cache_utils file:
https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/cache/http_cache_utils.cc

But I'm happy to fix it! Didn't know about that macro, thanks for pointing that out.

jmarantz commented on source/extensions/filters/http/cache/cacheability_utils.cc:
```diff
@@ -0,0 +1,60 @@
+
+namespace {
+
+bool static isCacheableStatusCode(absl::string_view status_code) {
```
s/static// (you are already in an anon namespace.

Also see above comment about not making these kinds of functions private so that you can unit-test them.

yosrym93 replied:
Will remove the static keyword.
This is also a function solely used and tested by another function (i.e isCacheableResponse). It was also @toddmgreer to move them to unnamed namespaces (which I agree with). What do you think Todd?

yosrym93 replied:
I will replace the function with a CONSTRUCT_ON_USE getter of the hash set.

jmarantz commented on source/extensions/filters/http/cache/cacheability_utils.cc:
```diff
@@ -0,0 +1,60 @@
+  // https://tools.ietf.org/html/rfc7725#section-3
+  // TODO(yosrym93): the list of cacheable status codes should be configurable
+  static const auto& cacheable_status_codes = *new absl::flat_hash_set<absl::string_view>{
+      "200", "203", "204", "206", "300", "301", "308", "404", "405", "410", "414", "451", "501"};
```
CONSTRUCT_ON_FIRST_USE ?

jmarantz commented on source/extensions/filters/http/cache/cacheability_utils.cc:
```diff
@@ -0,0 +1,60 @@
+} // namespace Cache
+} // namespace HttpFilters
+} // namespace Extensions
+} // namespace Envoy
```
add newline

jmarantz commented on source/extensions/filters/http/cache/http_cache.cc:
```diff
@@ -74,30 +73,63 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
+  SystemTime response_time = CacheHeadersUtils::httpTime(response_headers.Date());
+  SystemTime::duration response_age = timestamp_ - response_time;
+
+  ASSERT(response_age >= std::chrono::seconds(0), "Response time is in the future.");
```
Do a "<" comparison on timestamp_ vs response_time before doing the subtraction. And you are error-checking with an ASSERT. I'm OK with an assert but also return some kind of error status as well unless you can guarantee through code inspection this cannot occur.

The other alternative is a RELEASE_ASSERT. What is not OK is to fall past that in production code.

jmarantz commented on source/extensions/filters/http/cache/http_cache.cc:
```diff
@@ -74,30 +73,63 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
+  // is implemented, this ASSERT will need to be updated
+  ASSERT((response_headers.Date() && response_cache_control.max_age_.has_value()) ||
+             response_headers.get(Http::Headers::get().Expires),
+         "Cache entry does not have valid expiration data.");
```
again we are having error-checking with an ASSERT. We need to have a story for this behavior in production.

yosrym93 replied:
This is not error checking, CacheabilityUtils::isCacheableResponse makes sure no responses get cached unless they satisfy this assert.

jmarantz replied:
@htuch @asraa for policy clarification on this pattern of asserts.

htuch replied:
I agree with @yosrym93 here but it would be helpful to add comments to clarify who is providing the guarantee above each `ASSERT`.

jmarantz commented on test/extensions/filters/http/cache/cache_headers_utils_test.cc:
```diff
@@ -0,0 +1,291 @@
+        // Empty header
+        {
+          "", 
+          {false, false, false, false, UNSET_DURATION, UNSET_DURATION, UNSET_DURATION}
```
My main challenge here is that I have to keep scrolling back to the struct definition to see what you are trying to test. The other issue I have with this style of writing heavily parameterized tests is that if there is a failure it is pretty hard to suss out what case it was that failed. Certainly line-numbers printed with test failures will not help.

However I'm not going to get hung up over this stylistic issue for now. But at a minimum put comments in, at least every few test-cases, indicating the ordering, e.g.
```
  // must-validate   no-store.  no-transform  no-stale.  ....
```
actually it's hard to tell just from looking at this file what the 3 times even are.

yosrym93 replied:
I will add comments to try and make it more readable.
On failures, the ostream<< operator is implemented so the mismatch is printed properly.

For the meaning of the 3 durations they directly correspond to cache control directives with the same names, i.e max-age, min-fresh, and max-stale.

yosrym93 replied:
@jmarantz let me know if you can think of any more improvements for these tests readability.

jmarantz commented on test/extensions/filters/http/cache/cache_headers_utils_test.cc:
```diff
@@ -0,0 +1,320 @@
+public:
+  static auto getTestCases() {
+    // clang-format off
+    static const RequestCacheControlTestCase test_cases[] = {
```
does the TestResponseCacheControl actually get called? If so, this is a fiasco violation. If it doesn't, might as well just remove it.

See https://github.com/envoyproxy/envoy/blob/master/STYLE.md and search for 'fiasco' for context.

If this strictly used POD-initialization then it gets initialized at the compile time and there is no problem.

yosrym93 replied:
I will replace it with CONSTRUCT_ON_FIRST_USE and use a vector.

jmarantz commented on test/extensions/filters/http/cache/cache_headers_utils_test.cc:
```diff
@@ -0,0 +1,320 @@
+
+class RequestCacheControlTest : public testing::TestWithParam<RequestCacheControlTestCase> {
+public:
+  static auto getTestCases() {
```
use explicit type here

https://google.github.io/styleguide/cppguide.html#Type_deduction

jmarantz commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,66 @@
+  explicit RequestCacheControl(absl::string_view cache_control_header);
+
+  // must_validate is true if 'no-cache' directive is present
+  bool must_validate_ : 1;
```
ditto for bit-packing question.

yosrym93 commented:
/retest

junr03 commented:
@yosrym93 do you mind fixing format?

yosrym93 commented:
@junr03 It was the legacy codec problem, I merged master, it should be fixed.

junr03 commented:
Thanks, @htuch do you mind doing a pass on this (saw that @jmarantz mentioned you a couple times, so thought you could do the senior maintainer pass).

jmarantz commented:
Yes, in particular I want @htuch to look at the asserts.

htuch commented on source/extensions/filters/http/cache/cache_headers_utils.h:
```diff
@@ -0,0 +1,66 @@
+  bool no_stale_ = false;
+  // 'public' directive is not used now
+  bool is_public_ = false;
+  // max_age is set if to 's-maxage' if present, if not it is set to 'max-age' if present.
```
Can you summarize in comments what each field means semantically, it's a pain to keep looking at the RFC.

yosrym93 commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,171 @@
+// VCHAR           =  %x21-7E  ; visible (printing) characters
+
+RequestCacheControl::RequestCacheControl(absl::string_view cache_control_header) {
+  std::vector<absl::string_view> directives = absl::StrSplit(cache_control_header, ',');
```
I could add a link to https://httpwg.org/specs/rfc7234.html#collected.abnf which includes the format for this header and that multiple directives are separated by commas.

yosrym93 commented on source/extensions/filters/http/cache/cache_headers_utils.cc:
```diff
@@ -0,0 +1,171 @@
+  // Sun, 06 Nov 1994 08:49:37 GMT    ; IMF-fixdate
+  // Sunday, 06-Nov-94 08:49:37 GMT   ; obsolete RFC 850 format
+  // Sun Nov  6 08:49:37 1994         ; ANSI C's asctime() format
+  static const char* rfc7231_date_formats[] = {"%a, %d %b %Y %H:%M:%S GMT",
```
This function was originally written by @toddmgreer in `source/extensions/filters/http/cache/http_cache_utils.cc` , I just moved it to this file. I think he has the answer to this.

toddmgreer replied:
Compulsory.
See https://httpwg.org/specs/rfc7231.html#http.date, particularly the ABNF grammar.

htuch commented on source/extensions/filters/http/cache/http_cache.cc:
```diff
@@ -74,30 +73,67 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
-    return timestamp_ - HttpCacheUtils::httpTime(response_headers.Date()) < effective_max_age;
+
+  const SystemTime::duration response_age = timestamp_ - response_time;
+  bool request_max_age_exceeded = request_cache_control_.max_age_.has_value() &&
```
Nit: more const everywhere in this PR. There are a lot of local variables that can be made immutable, please update them across the files of the PR. Thanks.

yosrym93 commented:
@htuch I am fixing the const local variables now. For the coverage, the whole cache filter tree has low coverage before this PR (still WIP), this PR actually raises coverage by about 4-5% IIRC. I intend to start working on a separate PR to improve coverage & tests generally for this.
